api/PROFILE.md

```diff
@@ -99,6 +99,56 @@ The request MUST have the body of a single protobuf message of type [LocalAppInf
 Device publishes information repeatedly because of no expectation of persist storage inside local server. Device MAY
 throttle or cancel sending in case of Not implemented response.
 
+### Application Commands
+
+Retrieve a locally submitted commands targeted for application instances running
+on the device.
+
+GET /api/v1/app_commands
+
+Return codes:
+
+* Valid: `200`
+* Not implemented: `404`
+
+The request MUST use HTTP protocol.
+The request MUST NOT contain any body content.
+
+Response:
+
+The response mime type MUST be `application/x-proto-binary`. The response MUST contain
+a single protobuf message of type [LocalAppCmds](./proto/profile/local_profile.proto).
+
+The requester MUST verify that the response payload has the correct `server_token`.
+If the verification succeeds, all entries of `app_commands` are iterated, and those
+that successfully match a running application instance (by `uuid` and/or `displayname`)
+are applied.
+
+Currently, the method allows to request a locally running application instance
+to be *restarted* or *purged* by EVE. This may help to resolve a case of an application
+being in a broken state, and the user not being able to fix it (remotely) due to a lack
+of connectivity between the device and the controller. Rather than rebooting the entire
+device (locally), it is possible to restart/purge only a selected application.
+
+A command request, as defined by `AppCommand` protobuf message, includes an important
+field `timestamp` (`uint64`), which should record the time when the request was made
+by the user. The format of the timestamp is not defined. It can be a Unix timestamp
+or a different time representation. It is not even required for the timestamp to match
+the real time or to be in-sync with the device clock.
+
+What is required, however, is that two successive but distinct requests made for the same
+application and with the same command will have different timestamps attached.
+This requirement applies even between restarts of the Local profile server. A request made
+after a restart should not have the same timestamp attached as the previous request with
+the same arguments made before the restart.
+
+EVE guarantees that a newly added command request (into `LocalAppCmds.app_commands`),
+or a change of the `timestamp` field, will result in the command being triggered ASAP,
```

Comment:
Firstly, my understanding is that this functionality is needed to be able to fix app issues locally. So it is targeting a "live local troubleshooting". An eventual guarantee of a command being executed even if restart/reboot happens is probably unnecessary. Also I do not want to make things complicated on the Local profile server side by requiring command requests to be persisted.

So the idea was that EVE remembers (non-persistently) the timestamp of the last purge/restart/... for each application. The local profile server does not have to stop asserting the command even if it has been already executed. Unless the timestamp changes, EVE would not repeat the command. Reboot of the device (i.e. EVE forgetting the timestamps), also means that the local profile server is restarted and begins with an empty state.

Reply:
If neither remembers the state then if EVE restarts and the local profile server does not restart it will cause the command to happen again. I agree we don't want to make the local profile server more complex, but I do think it makes sense to have EVE remember the last timestamp for each app instance. (This is analogous to rembering the counters for the purge and restart commands comming from the controller in /persist/status/zedmanager/UuidToNum/.) FWIW the way we handle those numbers is that we do not update the recorded numbers until the restart or purge has completed to avoid a case where we hit a power failure after having received the new number in the API but before having acted on it.

Note that we might not need this persistent recording of the timestamp when the local-profile-server is deployed on EVE, but we currently do not require such a deployment pattern. Hence the local-profile-server can survive a power failure+restart of the device running EVE-OS.

Reply:
Oh, I see, I wrongly assumed that if EVE goes down then the local profile server also goes down. But what you are suggesting is that another component of the server might exists on another machine, which will survive and keep the state even if EVE restarts (?) Based on our API it seems that the local profile server could run even completely externally, but currently our implementation expects to find it among the apps deployed on the device.
In that case EVE really must do persistence. I will update the PR...